### PR TITLE
String solver

### PR DESCRIPTION
**Pull request by romainbrenguier**
This implements a string solver in CBMC. 
It is activated by the --pass option and the behaviour of CBMC should be unchanged otherwise.
It can be tested on the regression tests of the string directory.
This implements issues:
https://github.com/diffblue/verification-engine-utils/issues/146
https://github.com/diffblue/verification-engine-utils/issues/182
https://github.com/diffblue/verification-engine-utils/issues/223
https://github.com/diffblue/verification-engine-utils/issues/224
https://github.com/diffblue/verification-engine-utils/issues/227
https://github.com/diffblue/verification-engine-utils/issues/276
And replaces these pull requests that are now obsolete: 
https://github.com/diffblue/cbmc/pull/273
https://github.com/diffblue/cbmc/pull/276
https://github.com/diffblue/cbmc/pull/277
https://github.com/diffblue/cbmc/pull/278

### CODE REVIEW

File: src/solvers/refinement/string_constraint_generator.cpp [hunk 1]
```diff
@@ -0,0 +1,1840 @@
...
+{
+  string_exprt s1 = string_of_expr(args(f,2)[0]);
+  string_exprt s2 = string_of_expr(args(f,2)[1]);
+  typet return_type = f.type();
+  symbol_exprt res = string_exprt::fresh_symbol("compare_to",return_type);
+
+  // In the lexicographic comparison, x is the first point where the two strings differ.
+  // res == 0 => |s1| = |s2| && forall i < |s1|. s1[i] == s2[i]
+  // res != 0 =>
+  //   (|s1| <= |s2| && exists x < |s1|. res = s1[x] - s2[x] && forall i<x s
```
Comment (smowton) [inline src/solvers/refinement/string_constraint_generator.cpp]: I haven't read all of these since they're quite stereotyped, but the same fixes apply as elsewhere -- bring the coding style in line with the codebase standard, and name the functions informatively so that at callsites it's clear that they're creating something rather than just fetching some existing thing.

File: src/solvers/refinement/string_functions.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+const irep_idt cprover_string_parse_int_func("__CPROVER_uninterpreted_string_parse_int_func");
+const irep_idt cprover_string_replace_func("__CPROVER_uninterpreted_string_replace_func");
+const irep_idt cprover_string_set_length_func("__CPROVER_uninterpreted_string_set_length_func");
+const irep_idt cprover_string_startswith_func("__CPROVER_uninterpreted_string_startswith_func");
+const irep_idt cprover_string_substring_func("__CPROVER_uninterpreted_string_substring_func");
+const irep_idt cprover_string_to_char_array_func("__CPROVER_uninterpreted_string_to_char_array_func");
+const irep_idt cprover_string_to_lower_case_func("__CPROVER_uninterpreted_string_to_lower_case_func");
+const irep_idt cprover_string_to_upper_case_func("__CPROVER_uninterpreted_string_to_upper_case_func");
+const irep_idt cprover_string_trim_func("__CPROVER_uninterpreted_string_trim_func");
+const irep_idt cprover_string_value_of_func("__CPROVER_uninterpreted_string_value_of_func");
```
Comment (smowton) [inline src/solvers/refinement/string_functions.h]: Rather than define ireps like this, add to the lists built from util/irep*, creating ID_string_trim and so on. Consider replacing "uninterpreted" with just "string_solver" or similar, since the base language string functions aren't being interpreted either.

File: src/solvers/refinement/string_refinement.cpp [hunk 3]
```diff
@@ -0,0 +1,769 @@
...
+#include <langapi/languages.h>
+#include <ansi-c/ansi_c_language.h>
+
+// Succinct version of pretty()
+std::string string_refinementt::pretty_short(const exprt & expr) {
+  languagest languages(ns, new_ansi_c_language());
+  std::string string_value;
+  languages.from_expr(expr, string_value);
+  return string_value;
+}
```
Comment (smowton) [inline src/solvers/refinement/string_refinement.cpp]: This duplicates the global from_expr helper (see `src/langapi/language_util.h`)

File: src/solvers/refinement/string_refinement.cpp [hunk 4]
```diff
@@ -0,0 +1,769 @@
...
+    }
+
+  debug() << "adding lemma " << pretty_short(lemma) << eom;
+
+  prop.l_set_to_true(convert(lemma));
+  if(add_to_index_set)
+    cur.push_back(lemma);
+}
+
+unsigned integer_of_expr(const constant_exprt & expr)
```
Comment (smowton) [inline src/solvers/refinement/string_refinement.cpp]: Use arith-tools.h

File: src/solvers/refinement/string_refinement.cpp [hunk 5]
```diff
@@ -0,0 +1,769 @@
...
+public:
+  find_qvar_visitor(const exprt &qvar): qvar_(qvar) {}
+
+  void operator()(const exprt &expr) {
+    if (expr == qvar_) throw true;
+  }
+};
+
+// Look for the given symbol in the index expression
+bool find_qvar(const exprt index, const symbol_exprt & qvar)
```
Comment (smowton) [inline src/solvers/refinement/string_refinement.cpp]: If this is purely a local helper, make it `static`

File: src/solvers/refinement/refined_string_type.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+bool refined_string_typet::is_java_string_type(const typet &type)
+{
+  if(type.id() == ID_pointer) {
+    pointer_typet pt = to_pointer_type(type);
+    typet subtype = pt.subtype();
+    return is_java_deref_string_type(subtype);
+  } else return false;
+}
```
Comment (romainbrenguier) [inline src/solvers/refinement/refined_string_type.cpp]: No good reason for not catching it, I fixed that in https://github.com/diffblue/cbmc/pull/341/commits/ab4c970e450403b0354f12ae306a7cbf5b8ca120

File: src/solvers/refinement/string_constraint.cpp [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+string_constraintt string_constraintt::forall(const symbol_exprt & univ, const exprt & bound_sup)
+{
+  return forall(univ,refined_string_typet::index_zero(),bound_sup);
+}
+
+string_constraintt string_constraintt::not_contains(exprt univ_bound_inf, exprt univ_bound_sup,
+				 exprt premise, exprt exists_bound_inf,
+				 exprt exists_bound_sup, exprt s0, exprt s1)
+{
+  string_constraintt sc(premise);
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint.cpp]: Done in https://github.com/diffblue/cbmc/pull/341/commits/880812d989f9ee84c09d4bbc37a7863c668a4601
but this will probably end up being a constructor for its own class in another commit

File: src/solvers/refinement/string_constraint.h [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+  exprt premise() const;
+
+  exprt body() const;
+
+  inline exprt s0() const { assert(is_not_contains()); return compared_strings[0];}
+  inline exprt s1() const { assert(is_not_contains()); return compared_strings[1];}
+
+
+  inline symbol_exprt get_univ_var() const { assert(form==UNIV_QUANT); return quantified_variable;}
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint.h]: Ids introduced in https://github.com/diffblue/cbmc/pull/341/commits/0f6409cb740d5d2f9cab03cbd959e1b52a875a60

File: src/solvers/refinement/string_constraint.h [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+  // String axioms can have 4 different forms:
+  // either a simple expression p,
+  // or a string constant: forall x in [0,|s|[. s(x) = c(x)
+  // or universally quantified expression: forall x in [lb,ub[. p(x)
+  // or a expression for non containment:
+  // forall x in [lb,ub[. p(x) => exists y in [lb,ub[. s1[x+y] != s2[y]
+  enum {SIMPLE, STRING_CONSTANT, UNIV_QUANT, NOT_CONTAINS} form;
+
+  // Universally quantified symbol
+  symbol_exprt quantified_variable;
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint.h]: This has been fixed in https://github.com/diffblue/cbmc/pull/341/commits/0f6409cb740d5d2f9cab03cbd959e1b52a875a60

File: src/solvers/refinement/string_constraint_generator.cpp [hunk 10]
```diff
@@ -0,0 +1,1840 @@
...
+  else
+    if(language == UNKNOWN)
+      {
+	if(refined_string_typet::is_c_string_type(str.type()))
+	  language = C;
+	else
+	  language = JAVA;
+      }
+
+}
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator.cpp]: I did not find a very good way to do that. 
My attempt at making it cleaner is here https://github.com/diffblue/cbmc/pull/341/commits/3c1daf04eed37dffe7e3f062c5bdcd7b900ba443

File: src/solvers/refinement/string_constraint_generator.cpp [hunk 11]
```diff
@@ -0,0 +1,1840 @@
...
+
+symbol_exprt string_constraint_generatort::fresh_boolean(const irep_idt &prefix)
+{
+  symbol_exprt b = string_exprt::fresh_symbol(prefix,bool_typet());
+  boolean_symbols.push_back(b);
+  return b;
+}
+
+
+string_exprt string_constraint_generatort::string_of_expr(const exprt & unrefined_string)
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator.cpp]: Done in https://github.com/diffblue/cbmc/pull/341/commits/91ebe6198ca8f6d5d33c644f48eb914f0199f89e

File: src/solvers/refinement/string_constraint_generator.cpp [hunk 12]
```diff
@@ -0,0 +1,1840 @@
...
+string_exprt string_constraint_generatort::string_of_symbol(const symbol_exprt & sym)
+{ return get_string_of_symbol(sym); }
+
+
+exprt string_constraint_generatort::function_application
+(const function_application_exprt & expr)
+{
+  const exprt &name = expr.function();
+  assert(name.id() == ID_symbol);
+
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator.cpp]: In java, during the verification condition generation the ids gets some suffix added (for instance we get `__CPROVER_uninterpreted_string_of_int_func#0`) so I needed starts_with here. 
If there is a way to avoid getting this suffix then it would be a better to use equality.

File: src/solvers/refinement/string_constraint_generator.cpp [hunk 13]
```diff
@@ -0,0 +1,1840 @@
...
+    }
+}
+
+
+irep_idt extract_java_string(const symbol_exprt & s)
+{
+  std::string tmp(s.get(ID_identifier).c_str());
+  std::string value = tmp.substr(31);
+  return irep_idt(value);
+}
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator.cpp]: It will be when the better representation gets into CBMC but as far as I know, it's not yet the case.
We can either let it like that and make another pull request with the new representation and fixing this hack; or merge a pull request with the new representation before this one.

File: src/solvers/refinement/string_constraint_generator.cpp [hunk 14]
```diff
@@ -0,0 +1,1840 @@
...
+  string_exprt res(char_type);
+  std::string str = sval.c_str();
+  // should only do this for java
+  std::wstring utf16 = utf8_to_utf16le(str);
+
+  for (std::size_t i = 0; i < utf16.size(); ++i) {
+    std::string idx_binary = integer2binary(i,STRING_SOLVER_INDEX_WIDTH);
+    constant_exprt idx(idx_binary, refined_string_typet::index_type());
+    std::string sval_binary=integer2binary((unsigned)utf16[i], char_width);
+    constant_exprt c(sval_binary,char_type);
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator.cpp]: Done in https://github.com/diffblue/cbmc/pull/341/commits/4c5d3526d4a2644b2592693cf2f0e927be0fb0fb

File: src/solvers/refinement/string_functions.cpp [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+#include <solvers/refinement/string_functions.h>
+
+bool starts_with(irep_idt id, irep_idt prefix) {
+  std::string s = id2string(id);
+  std::string t = id2string(prefix);
+  for(int i = 0; i < t.length(); i++)
+    if(s[i] != t[i]) return false;
+  return true;
+}
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_functions.cpp]: done in https://github.com/diffblue/cbmc/pull/341/commits/9efef8fd31d0b336b082ee7b754a0c6b2dd9d496

File: src/cbmc/cbmc_parse_options.cpp [hunk 16]
```diff
@@ -1188,6 +1201,7 @@ void cbmc_parse_optionst::help()
     " --yices                      use Yices\n"
     " --z3                         use Z3\n"
     " --refine                     use refinement procedure (experimental)\n"
+    " --pass                       use parameterized array for string solving (experimental)\n"
```
Comment (kroening) [inline src/cbmc/cbmc_parse_options.cpp]: Please replace by --refine-string; the name "pass" isn't useful for anybody.

**Comment by kroening** [comment]
Please replace the file names "pass" by "string_refine"; the acronym isn't widely known.

File: src/solvers/refinement/string_constraint_generator.cpp (hunk 12 above)
Comment (tautschnig) [inline src/solvers/refinement/string_constraint_generator.cpp]: I believe the following should do:
```
is_ssa_expr(expr) ?
to_ssa_expr(expr).get_object_name() :
to_symbol_expr(expr).get_identifier()
```
to get rid of "#0"

File: src/solvers/refinement/string_constraint_generator.cpp (hunk 13 above)
Comment (tautschnig) [inline src/solvers/refinement/string_constraint_generator.cpp]: Seeing `substr(31)`  (fixed offset) is really scary...

File: src/solvers/refinement/string_functions.h (hunk 2 above)
Comment (romainbrenguier) [inline src/solvers/refinement/string_functions.h]: This has been done in https://github.com/diffblue/cbmc/pull/341/commits/c81641afc803acc0011bc32ef84a969c282f7173
The string_functions.h file is now useless and has been removed.
This change also made necessary to deal with C string functions in a different way. They are now processed by pass_preprocess in a similar way to java functions

File: src/cbmc/cbmc_parse_options.cpp (hunk 16 above)
Comment (romainbrenguier) [inline src/cbmc/cbmc_parse_options.cpp]: Replaced it https://github.com/diffblue/cbmc/pull/341/commits/31b20a554f0c6e09917747a44e8a54e418e458f5

File: src/solvers/refinement/string_refinement.cpp (hunk 3 above)
Comment (romainbrenguier) [inline src/solvers/refinement/string_refinement.cpp]: Used `from_expr` in that commit: https://github.com/diffblue/cbmc/pull/341/commits/d39d0142451c725dff0d4c50d9e8975e2607352d

File: src/solvers/refinement/string_refinement.cpp (hunk 4 above)
Comment (romainbrenguier) [inline src/solvers/refinement/string_refinement.cpp]: Done in https://github.com/diffblue/cbmc/pull/341/commits/2ab3e49e3a552a59eab3c1b54e104f3d49161f93

File: src/config.inc [hunk 17]
```diff
@@ -3,6 +3,7 @@ BUILD_ENV = AUTO
 
 # Variables you may want to override
 #CXXFLAGS = -Wall -O0 -g -Werror -Wno-long-long -Wno-sign-compare -Wno-parentheses -Wno-strict-aliasing -pedantic
+CXXFLAGS = -Wall -O0 -g -Wno-long-long -Wno-sign-compare -Wno-parentheses -Wno-strict-aliasing -pedantic
```
Comment (kroening) [inline src/config.inc]: Please remove this change.

Comment (romainbrenguier) [inline src/config.inc]: yes sorry for pushing that, I removed the change from the commit

File: src/solvers/refinement/string_constraint_generator.cpp (hunk 13 above)
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator.cpp]: I've put a slightly cleaner version of this function in this commit https://github.com/diffblue/cbmc/pull/341/commits/2e7f98a2cbd108ecfad4f0a560f805367c8223b2

**Comment by romainbrenguier** [comment]
Regarding the coding standards, I fixed the problems and checked the files with the cpplint script, https://github.com/diffblue/cbmc/pull/341/commits/d6d4627ff9990d6199ab3b4726c18ceb0bec3ad9

**Comment by romainbrenguier** [comment]
@tautschnig yes I think I will split this into several pull requests, one for each of these points:
* changes to the regression tests
* unicode functions
* the string solver itself
* the string preprocessing for goto-programs
* the option `--string-refine` added to CBMC

and close this one when the requests are made
